### PR TITLE
enhancement(vrl): Add VRL function get_vector_timezone

### PR DESCRIPTION
**Pull request by klondikedragon**
closes: https://github.com/vectordotdev/vector/issues/19716

This function returns the name of the configured timezone, or if set to `local` then it will attempt to get the timezone name from the host OS. If that fails, then it will get the local time, and return the formatted timezone offset (e.g., "+02:00").

This has been tested with and without a configured timezone, and also tested the condition when the call to get the local timezone name fails and it has to produce an offset instead.

I initially thought that this new VRL function would be specific to vector rather than VRL stdlib, since it needed the value of vector's global configuration `timezone` value. However, as part of developing this PR, I learned that the timezone is a part of the core VRL `Context` and that vector passes in the value of its global configuration `timezone` value [here](https://github.com/vectordotdev/vector/blob/master/src/transforms/remap.rs#L440).

This means that probably this function should be implemented in the [VRL stdlib](https://github.com/vectordotdev/vrl/tree/main/src/stdlib) instead? If so, probably the name should then be something like `get_timezone_name()`?

I'm submitting the PR here for discussion, and then to ask, in order to add this to VRL stdlib, would the procedure be something like the following?
1. Create PR to core VRL stdlib
2. Wait for VRL stdlib with enhancement to be released
3. Update this PR to reflect new version of vrl stdlib (and to change the documentation to reflect the new more generic function name, etc.)

What do you think?

### CODE REVIEW

**Comment by klondikedragon** [comment]
I forgot to mention, this resolves #19716

**Comment by pront** [comment]
Also, some checks are failings. Use `cue fmt` and `make check-component-docs`.

**Comment by klondikedragon** [comment]
@pront - should be OK now, see what you think; I had run local tests & doc checks before, but apparently local checks skip cue format check. I had to run `CI=true make check-docs` locally to reproduce cue format issue. Also, it looks like changelog.d fragments are fairly new, and spell checker doesn't like arbitrary names in authors line, so I updated the spellcheck patterns file to ignore those lines. Thanks!

File: website/cue/reference/remap/functions/get_vector_timezone.cue [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package metadata
+
+remap: functions: get_vector_timezone: {
+	category:    "System"
+	description: """
+		Returns the name of the timezone in the vector configuration
```
Comment (maycmlee) [inline website/cue/reference/remap/functions/get_vector_timezone.cue:6]:
```suggestion
		Returns the name of the timezone in the Vector configuration.
```

File: website/cue/reference/remap/functions/get_vector_timezone.cue [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+		(see [global configuration options](\(urls.vector_configuration_global))).
+		If the configuration is set to `local`, then it attempts to
+		determine from the host OS the name of the timezone. If this
+		is not possible, then it will return the fixed offset of the
```
Comment (maycmlee) [inline website/cue/reference/remap/functions/get_vector_timezone.cue:10]:
```suggestion
		is not possible, then it returns the fixed offset of the
```

File: website/cue/reference/remap/functions/get_vector_timezone.cue [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+		local timezone for the current time in the format `"[+-]HH:MM"`,
+		for example, `"+02:00"`
```
Comment (maycmlee) [inline website/cue/reference/remap/functions/get_vector_timezone.cue:12]:
```suggestion
		for example, `"+02:00"`.
```

File: website/cue/reference/remap/functions/get_vector_timezone.cue [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+	arguments: []
+	internal_failure_reasons: [
+		"Retrieval of local timezone information failed.",
+	]
+	return: types: ["string"]
+
+	examples: [
+		{
+			title: "Get vector timezone when timezone config is set to 'America/Chicago'"
```
Comment (maycmlee) [inline website/cue/reference/remap/functions/get_vector_timezone.cue:23]:
```suggestion
			title: "Get Vector's timezone when the timezone config is set to 'America/Chicago'"
```

File: website/cue/reference/remap/functions/get_vector_timezone.cue [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+		{
+			title: "Get vector timezone when timezone config is set to 'local' and the host OS has a local timezone of America/New_York"
+			input: log: {}
+			source: #"""
+				.vector_timezone = get_vector_timezone!()
+				"""#
+			output: log: vector_timezone: "America/New_York"
+		},
+		{
+			title: "Get vector timezone when timezone config is set to 'local' and the host OS can only determine the local timezone offset"
```
Comment (maycmlee) [inline website/cue/reference/remap/functions/get_vector_timezone.cue:39]:
```suggestion
			title: "Get Vector's timezone when the timezone config is set to 'local' and the host OS can only determine the local timezone offset"
```

**Review by maycmlee — COMMENTED** [review]
Left some suggestions!

File: .github/actions/spelling/patterns.txt [hunk 6]
```diff
@@ -224,3 +224,6 @@ user:P@ssw0rd
 
 # Ignore base64 encoded values in Prometheus Pushgateway URL paths
 /.+@base64/.+
+
+# changelog.d fragment authors line
```
Comment (pront) [inline .github/actions/spelling/patterns.txt:228]: cc @neuronull

Comment (pront) [reply .github/actions/spelling/patterns.txt:228]: Thank you for checking!

@klondikedragon there's a few doc suggestions, after that this should be ready to merge.

File: website/cue/reference/remap/functions/get_vector_timezone.cue (hunk 1 above)
Comment (klondikedragon) [reply website/cue/reference/remap/functions/get_vector_timezone.cue:6]: I've changed to `Vector` -- note the parenthetical continuation of the sentence on the next line, so I kept the period after the clause in parens, see what you think.

File: website/cue/reference/remap/functions/get_vector_timezone.cue [hunk 7]
```diff
@@ -0,0 +1,47 @@
+package metadata
+
+remap: functions: get_vector_timezone: {
+	category:    "System"
+	description: """
+		Returns the name of the timezone in the vector configuration
+		(see [global configuration options](\(urls.vector_configuration_global))).
+		If the configuration is set to `local`, then it attempts to
+		determine from the host OS the name of the timezone. If this
```
[the remark this answers is not in the document]
Comment (klondikedragon) [reply website/cue/reference/remap/functions/get_vector_timezone.cue:9]: reads better, thanks!

File: .github/actions/spelling/patterns.txt (hunk 6 above)
Comment (klondikedragon) [reply .github/actions/spelling/patterns.txt:228]: thanks @maycmlee - @pront should be ready for another check

**Comment by jszwedko** [comment]
Thanks for this @klondikedragon !

Coming back to this question:

> This means that probably this function should be implemented in the [VRL stdlib](https://github.com/vectordotdev/vrl/tree/main/src/stdlib) instead? If so, probably the name should then be something like get_timezone_name()?

I tend to agree. Since the VRL runtime expects a timezone to be passed in, I think this function could be implemented in VRL instead. What would you think of moving this over there? cc/ @pront for any opposing thoughts.

**Comment by pront** [comment]
> Thanks for this @klondikedragon !
> 
> Coming back to this question:
> 
> > This means that probably this function should be implemented in the [VRL stdlib](https://github.com/vectordotdev/vrl/tree/main/src/stdlib) instead? If so, probably the name should then be something like get_timezone_name()?
> 
> I tend to agree. Since the VRL runtime expects a timezone to be passed in, I think this function could be implemented in VRL instead. What would you think of moving this over there? cc/ @pront for any opposing thoughts.

No objections at all. SGTM.

**Comment by klondikedragon** [comment]
@jszwedko @pront - I've opened https://github.com/vectordotdev/vrl/pull/671 - once that is merged and a VRL version is released with the update, I'll update this PR to incorporate and adapt the docs, etc.

**Comment by klondikedragon** [comment]
@jszwedko with https://github.com/vectordotdev/vrl/pull/671 now merged, what's the right process? should I wait until a new VRL is released so that the dependency here can be updated to the new version first? or should I update this PR just to add the appropriate doc and versions/dependencies will be worked out prior to release? Thanks!

**Comment by klondikedragon** [comment]
@jszwedko - this PR was updated to just be the documentation and test. The test will fail until this PR is rebased against a main branch that depends on the new VRL lib, etc.

**Comment by klondikedragon** [comment]
@jszwedko - rebased, should be good to go now

**Review by jszwedko — APPROVED** [review]
Perfect, thanks!

File: website/cue/reference/remap/functions/get_timezone_name.cue [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+			output: log: vector_timezone: "America/New_York"
+		},
+		{
+			title: "Get Vector's timezone when the timezone config is set to 'local' and the host OS can only determine the local timezone offset"
+			input: log: {}
+			source: #"""
+				.vector_timezone = get_timezone_name!()
+				"""#
+			output: log: vector_timezone: "-05:00"
+		},
```
Comment (jszwedko) [inline website/cue/reference/remap/functions/get_timezone_name.cue:45]:
```suggestion
```

File: website/cue/reference/remap/functions/get_timezone_name.cue [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	return: types: ["string"]
+
+	examples: [
+		{
+			title: "Get Vector's timezone when the timezone config is set to 'America/Chicago'"
+			input: log: {}
+			source: #"""
+				.vector_timezone = get_timezone_name!()
+				"""#
+			output: log: vector_timezone: "America/Chicago"
```
Comment (jszwedko) [inline website/cue/reference/remap/functions/get_timezone_name.cue:28]:
```suggestion
			output: log: vector_timezone: "UTC"
```

These examples run in CI which is in UTC.

File: changelog.d/19716_doc_vrl_get_timezone_name.enhancement.md [hunk 10]
```diff
@@ -0,0 +1,3 @@
+New VRL function `get_timezone_name` to return the configured/resolved IANA timezone name.
```
Comment (pront) [inline changelog.d/19716_doc_vrl_get_timezone_name.enhancement.md:1]: Do we need this fragment?

Comment (jszwedko) [reply changelog.d/19716_doc_vrl_get_timezone_name.enhancement.md:1]: Ah, no, we don't since it'll pull in the fragment from VRL.

Comment (jszwedko) [inline changelog.d/19716_doc_vrl_get_timezone_name.enhancement.md:1]: Could we drop this file @klondikedragon ? It'll actually pull in the VRL changelog entry when we generate the Vector changelog so we don't need this one too.

**Comment by jszwedko** [comment]
(I tried to make these changes myself but appear to lack access to[ push to the fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork))

File: changelog.d/19716_doc_vrl_get_timezone_name.enhancement.md (hunk 10 above)
Comment (klondikedragon) [reply changelog.d/19716_doc_vrl_get_timezone_name.enhancement.md]: no problem, file deleted

File: website/cue/reference/remap/functions/get_timezone_name.cue (hunk 8 above)
Comment (klondikedragon) [reply website/cue/reference/remap/functions/get_timezone_name.cue]: removed

File: website/cue/reference/remap/functions/get_timezone_name.cue (hunk 9 above)
Comment (klondikedragon) [reply website/cue/reference/remap/functions/get_timezone_name.cue]: updated

**Comment by klondikedragon** [comment]
@jszwedko - changes made. I'd be happy for you to make the changes too :) -- it looks like that github feature only works if the fork is from a user rather than an org like it is in this case. hopefully these changes are what's needed to get it done.

**Comment by jszwedko** [comment]
Thanks @klondikedragon ! I marked this to merge.

**Pull request merged**